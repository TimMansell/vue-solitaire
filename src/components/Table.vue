<template>
  <table class="table" :class="tableClass">
    <thead>
      <tr data-test="table-header-row">
        <th
          class="table__cell"
          v-for="(heading, headingIndex) in headings"
          :key="headingIndex"
        >
          {{ heading }}
        </th>
      </tr>
    </thead>
    <tbody v-if="!rows.length">
      <tr
        v-for="(row, placeholderRowIndex) in placeholderRows"
        :key="placeholderRowIndex"
        data-test="table-placeholder-row"
      >
        <td
          class="table__cell"
          v-for="(cell, placeholderCellIndex) in headings"
          :key="placeholderCellIndex"
        >
          <SkeletonLoader />
        </td>
      </tr>
    </tbody>
    <tbody v-if="rows.length">
      <tr
        class="table__row"
        :class="{ 'table__row--highlighted': isHighlighted }"
        v-for="({ row, isHighlighted }, rowIndex) in rows"
        :key="rowIndex"
        data-test="table-row"
      >
        <td
          class="table__cell"
          v-for="(cell, cellIndex) in row"
          :key="cellIndex"
          data-test="table-cell"
        >
          {{ cell }}
        </td>
      </tr>
    </tbody>
  </table>
</template>

<script>
import SkeletonLoader from '@/components/SkeletonLoader.vue';

export const isRowHighlighted = (cells, toHighlight) => {
  if (toHighlight) {
    const { key, value } = toHighlight;

    return cells[key] === value;
  }

  return false;
};

export default {
  name: 'Table',
  components: {
    SkeletonLoader,
  },
  props: {
    headings: {
      type: Array,
      default: () => [],
    },
    items: {
      type: Array,
      default: () => [],
    },
    spacing: {
      type: Boolean,
      default: false,
    },
    placeholderRows: {
      type: Number,
      default: 1,
    },
    toHighlight: {
      type: Object,
      default: () => {},
    },
  },
  computed: {
    tableClass() {
      const { spacing } = this;

      return {
        'table--spacing': spacing,
      };
    },
    rows() {
      const { items, toHighlight } = this;

      const formatteditems = items.map((cells) => ({
        row: { ...cells },
        isHighlighted: isRowHighlighted(cells, toHighlight),
      }));

      return formatteditems;
    },
  },
};
</script>

<style scoped lang="scss">
.table {
  width: 100%;
  background: var(--bg-primary-alt-2);

  &--spacing {
    margin-bottom: var(--mg-md);
  }

  &__row {
    background: transparent;

    &--highlighted {
      background: var(--bg-tertiary-alt);
    }
  }

  &__cell {
    border: 1px solid var(--bdr-secondary);
    padding: var(--pd-sm);
    white-space: nowrap;
    line-height: 1;
  }
}
</style>
