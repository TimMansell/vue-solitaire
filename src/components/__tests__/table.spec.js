import { shallowMount } from '@vue/test-utils';
import Table, { isRowHighlighted } from '@/components/Table.vue';

const propsData = {
  headings: ['Heading 1', 'Heading 2', 'Heading 3'],
  items: [
    ['Cell 1', 'Cell 2', 'Cell 3'],
    ['Cell 4', 'Cell 5', 'Cell 6'],
  ],
};

describe('Table.vue', () => {
  it('matches snapshot', () => {
    const wrapper = shallowMount(Table, {
      propsData,
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('matches snapshot - placeholder rows', () => {
    const wrapper = shallowMount(Table, {
      propsData: {
        ...propsData,
        items: [],
      },
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should add spacing class', () => {
    const wrapper = shallowMount(Table, {
      propsData: {
        ...propsData,
        spacing: true,
      },
    });

    expect(wrapper.classes()).toContain('table--spacing');
  });

  it('should populate table headers correctly', () => {
    const wrapper = shallowMount(Table, {
      propsData,
    });

    expect(
      wrapper.findAll('[data-test="table-header-cell"]').at(0).text()
    ).toBe(propsData.headings[0]);

    expect(
      wrapper.findAll('[data-test="table-header-cell"]').at(1).text()
    ).toBe(propsData.headings[1]);

    expect(
      wrapper.findAll('[data-test="table-header-cell"]').at(2).text()
    ).toBe(propsData.headings[2]);
  });

  it('should populate table body cells correctly', () => {
    const wrapper = shallowMount(Table, {
      propsData,
    });

    expect(wrapper.findAll('[data-test="table-cell"]').at(0).text()).toBe(
      propsData.items[0][0]
    );

    expect(wrapper.findAll('[data-test="table-cell"]').at(1).text()).toBe(
      propsData.items[0][1]
    );

    expect(wrapper.findAll('[data-test="table-cell"]').at(2).text()).toBe(
      propsData.items[0][2]
    );

    expect(wrapper.findAll('[data-test="table-cell"]').at(3).text()).toBe(
      propsData.items[1][0]
    );

    expect(wrapper.findAll('[data-test="table-cell"]').at(4).text()).toBe(
      propsData.items[1][1]
    );

    expect(wrapper.findAll('[data-test="table-cell"]').at(5).text()).toBe(
      propsData.items[1][2]
    );
  });

  it('should return highlighted row', () => {
    const result = isRowHighlighted([1, 2, 3], { key: '0', value: 1 });

    expect(result).toBe(true);
  });

  it('should not return highlighted row', () => {
    const result = isRowHighlighted([1, 2, 3], { key: '0', value: 2 });

    expect(result).toBe(false);
  });

  it('should show placeholder rows', () => {
    const wrapper = shallowMount(Table, {
      propsData: {
        ...propsData,
        items: [],
      },
    });

    expect(wrapper.find('[data-test="table-placeholder-row"]').exists()).toBe(
      true
    );

    expect(wrapper.find('[data-test="table-row"]').exists()).toBe(false);
  });

  it('should not show placeholder rows', () => {
    const wrapper = shallowMount(Table, {
      propsData,
    });

    expect(wrapper.find('[data-test="table-placeholder-row"]').exists()).toBe(
      false
    );

    expect(wrapper.find('[data-test="table-row"]').exists()).toBe(true);
  });

  it('should show highlight correct row', () => {
    const wrapper = shallowMount(Table, {
      propsData: {
        ...propsData,
        toHighlight: { key: '0', value: 'Cell 1' },
      },
    });

    expect(
      wrapper.findAll('[data-test="table-row"]').at(0).classes()
    ).toContain('table__row--highlighted');
  });
});
