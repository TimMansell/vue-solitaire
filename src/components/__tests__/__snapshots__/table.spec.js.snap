// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table.vue matches snapshot - placeholder rows 1`] = `
<table class="table">
  <thead>
    <tr data-test="table-header-row">
      <th data-test="table-header-cell" class="table__cell">
        Heading 1
      </th>
      <th data-test="table-header-cell" class="table__cell">
        Heading 2
      </th>
      <th data-test="table-header-cell" class="table__cell">
        Heading 3
      </th>
    </tr>
  </thead>
  <tbody>
    <tr data-test="table-placeholder-row">
      <td class="table__cell">
        <skeletonloader-stub></skeletonloader-stub>
      </td>
      <td class="table__cell">
        <skeletonloader-stub></skeletonloader-stub>
      </td>
      <td class="table__cell">
        <skeletonloader-stub></skeletonloader-stub>
      </td>
    </tr>
  </tbody>
  <!---->
</table>
`;

exports[`Table.vue matches snapshot 1`] = `
<table class="table">
  <thead>
    <tr data-test="table-header-row">
      <th data-test="table-header-cell" class="table__cell">
        Heading 1
      </th>
      <th data-test="table-header-cell" class="table__cell">
        Heading 2
      </th>
      <th data-test="table-header-cell" class="table__cell">
        Heading 3
      </th>
    </tr>
  </thead>
  <!---->
  <tbody>
    <tr data-test="table-row" class="table__row">
      <td data-test="table-cell" class="table__cell">
        Cell 1
      </td>
      <td data-test="table-cell" class="table__cell">
        Cell 2
      </td>
      <td data-test="table-cell" class="table__cell">
        Cell 3
      </td>
    </tr>
    <tr data-test="table-row" class="table__row">
      <td data-test="table-cell" class="table__cell">
        Cell 4
      </td>
      <td data-test="table-cell" class="table__cell">
        Cell 5
      </td>
      <td data-test="table-cell" class="table__cell">
        Cell 6
      </td>
    </tr>
  </tbody>
</table>
`;
